### PR TITLE
[GEP-28] Handle networking and control plane components

### PR DESCRIPTION
**Pull request by rfranzke**
**How to categorize this PR?**

/area ipcei
/kind enhancement

**What this PR does / why we need it**:
This PR is the next increment for `gardenadm init`. It deploys `kube-proxy`, the `Network` extension, CoreDNS, the `NetworkPolicy`s, and finally re-deploys `gardener-resource-manager` and extension controllers into the pod network.

In addition, the `ControlPlane` resource is deployed as well as `Deployment`s/`StatefulSet` for the control plane components that have to be translated to static pods. These resources are used to make extension webhooks work (i.e., still allow extensions to modify the control plane components).

Finally, the `gardener-node-agent` systemd unit is activated, and the node is properly tainted with the common `node-role.kubernetes.io/control-plane` taint.

**Which issue(s) this PR fixes**:
Part of https://github.com/gardener/gardener/issues/2906

**Special notes for your reviewer**:
/cc @ScheererJ 
~~Still in draft because some unit tests are still missing.~~

**Release note**:

```other operator
NONE
```

### CODE REVIEW

File: pkg/gardenadm/botanist/extensions_test.go [hunk 1]
```diff
@@ -135,4 +147,89 @@ var _ = Describe("Extensions", func() {
...
+		})
+
+		It("should fail if a ManagedResource is not healthy", func() {
+			Expect(fakeClient.Create(ctx, managedResource1)).To(Succeed())
+			Expect(fakeClient.Create(ctx, managedResource2)).To(Succeed())
+
+			Expect(b.WaitUntilExtensionControllerInstallationsHealthy(ctx)).To(MatchError(ContainSubstring("is not healthy")))
+		})
+
+		It("should fail if all ManagedResource are healthy", func() {
```
Comment (ScheererJ) [inline pkg/gardenadm/botanist/extensions_test.go]:
```
		It("should succeed if all ManagedResource are healthy", func() {
```

File: extensions/pkg/webhook/cmd/options.go [hunk 2]
```diff
@@ -294,6 +303,45 @@ func (c *AddToManagerConfig) AddToManager(ctx context.Context, mgr manager.Manag
...
 
+	if mergeShootWebhooksIntoSeedWebhooks {
+		clientConfig := func(in admissionregistrationv1.WebhookClientConfig) (admissionregistrationv1.WebhookClientConfig, error) {
+			var path string
+			if in.Service != nil {
+				path = ptr.Deref(in.Service.Path, "")
+			} else if u := in.URL; u != nil {
+				parsedURL, err := url.Parse(*u)
+				if err != nil {
+					return admissionregistrationv1.WebhookClientConfig{}, fmt.Errorf("failed to parse URL: %v", err)
```
Comment (ScheererJ) [inline extensions/pkg/webhook/cmd/options.go]: Not sure if the error will also include the URL or if it should be added here for context, e.g.

```suggestion
					return admissionregistrationv1.WebhookClientConfig{}, fmt.Errorf("failed to parse URL %q: %w", *u, err)
```

WDYT?

File: charts/gardener/provider-local/templates/deployment.yaml [hunk 3]
```diff
@@ -90,6 +90,9 @@ spec:
         {{- if .Values.gardener.version }}
         - --gardener-version={{ .Values.gardener.version }}
         {{- end }}
+        - {{- if .Values.gardener.autonomousShootCluster }}
+        - --autonomous-shoot-cluster={{ .Values.gardener.autonomousShootCluster }}
+        - {{- end }}
```
Comment (ScheererJ) [inline charts/gardener/provider-local/templates/deployment.yaml]: No need for empty command line parameters:

```suggestion
        {{- if .Values.gardener.autonomousShootCluster }}
        - --autonomous-shoot-cluster={{ .Values.gardener.autonomousShootCluster }}
        {{- end }}
```

File: pkg/gardenadm/cmd/init/init.go [hunk 4]
```diff
@@ -148,6 +150,18 @@ func run(ctx context.Context, opts *Options) error {
...
+			Fn:           b.DeployKubeProxy,
+			SkipIf:       !kubeProxyEnabled,
+			Dependencies: flow.NewTaskIDs(waitUntilGardenerResourceManagerReady, waitUntilShootNamespacesReady, waitForExtensionControllersReady),
+		})
+		_ = g.Add(flow.Task{
+			Name:         "Deleting kube-proxy system component",
+			Fn:           b.Shoot.Components.SystemComponents.KubeProxy.Destroy,
+			SkipIf:       kubeProxyEnabled,
+			Dependencies: flow.NewTaskIDs(waitUntilGardenerResourceManagerReady),
+		})
```
Comment (ScheererJ) [inline pkg/gardenadm/cmd/init/init.go]: Is this just being a good citizen and including the clean-up as well or is there more to it? I would have hoped nobody changes there mind about `kube-proxy` deployment half-way through `gardenadm init`. Day 2 lifecycle management should not be performed by `gardenadm`.
That being said, I do not mind this single task and was just wondering if we changed directions meanwhile.

File: pkg/gardenadm/botanist/network.go [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	resourcesv1alpha1 "github.com/gardener/gardener/pkg/apis/resources/v1alpha1"
+	"github.com/gardener/gardener/pkg/component/networking/coredns"
+	"github.com/gardener/gardener/pkg/utils/kubernetes/health"
+)
+
+// IsPodNetworkAvailable checks if the ManagedResource for CoreDNS is deployed and ready. If yes, pod network must be
+// available (otherwise, CoreDNS (which runs in this network) wouldn't be available).
+func (b *AutonomousBotanist) IsPodNetworkAvailable(ctx context.Context) (bool, error) {
```
Comment (ScheererJ) [inline pkg/gardenadm/botanist/network.go]: Nit:

```suggestion
// IsPodNetworkAvailable checks if the ManagedResource for CoreDNS is deployed and ready. If yes, pod network must be
// available. Otherwise, CoreDNS which runs in this network wouldn't be available.
func (b *AutonomousBotanist) IsPodNetworkAvailable(ctx context.Context) (bool, error) {
```

File: pkg/gardenadm/botanist/network_test.go [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+		BeforeEach(func() {
+			managedResource = &resourcesv1alpha1.ManagedResource{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "shoot-core-coredns",
+					Namespace: namespace,
+				},
+			}
+		})
+
+		It("should return false because the ManagedResource does not exist", func() {
```
Comment (ScheererJ) [inline pkg/gardenadm/botanist/network_test.go:79]: A relevant scenario, which is handled in the code, but not in any test is the case that the `ManagedResource` crd has not yet been deployed. Should we add a test for it as well? WDYT?

File: test/e2e/gardenadm/hightouch/gardenadm.go [hunk 7]
```diff
@@ -141,6 +141,41 @@ var _ = Describe("gardenadm high-touch scenario tests", Label("gardenadm", "high
...
 		}, SpecTimeout(time.Minute))
 
+		It("should ensure the control plane namespace is properly labeled", func(ctx SpecContext) {
+			Eventually(ctx, func(g Gomega) map[string]string {
+				namespace := &corev1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: "kube-system"}}
+				g.Expect(shootClientSet.Client().Get(ctx, client.ObjectKeyFromObject(namespace), namespace)).To(Succeed())
+				return namespace.Labels
+			}).Should(HaveKeyWithValue("gardener.cloud/role", "shoot"))
+		}, SpecTimeout(time.Minute))
+
```
Comment (ScheererJ) [inline test/e2e/gardenadm/hightouch/gardenadm.go:142]: This seems to be duplicated somehow (see above).

```suggestion
```

File: pkg/gardenadm/botanist/network.go [hunk 8]
```diff
@@ -30,3 +37,42 @@ func (b *AutonomousBotanist) IsPodNetworkAvailable(ctx context.Context) (bool, e
...
+
+	namespaceList := &corev1.NamespaceList{}
+	if err := b.SeedClientSet.Client().List(ctx, namespaceList); err != nil {
+		return fmt.Errorf("failed listing namespaces: %w", err)
+
+	}
+	for _, namespace := range namespaceList.Items {
+		b.Logger.Info("Reconciling NetworkPolicies using gardenlet's reconciliation logic", "namespaceName", namespace.Name)
+		if _, err := reconciler.Reconcile(reconcilerCtx, reconcile.Request{NamespacedName: types.NamespacedName{Name: namespace.Name}}); err != nil {
+			return fmt.Errorf("failed running NetworkPolicy controller for %q: %w", namespace.Name, err)
```
Comment (ScheererJ) [inline pkg/gardenadm/botanist/network.go]:
```suggestion
			return fmt.Errorf("failed running NetworkPolicy controller for namespace %q: %w", namespace.Name, err)
```

File: pkg/gardenadm/cmd/init/init.go [hunk 9]
```diff
@@ -143,6 +143,11 @@ func run(ctx context.Context, opts *Options) error {
 			Fn:           b.WaitUntilExtensionControllerInstallationsHealthy,
 			Dependencies: flow.NewTaskIDs(deployExtensionControllers),
 		})
+		deployNetworkPolicies = g.Add(flow.Task{
+			Name:         "Deploying network policies",
+			Fn:           b.ApplyNetworkPolicies,
+			Dependencies: flow.NewTaskIDs(deployGardenerResourceManager, deployExtensionControllers),
+		})
```
Comment (ScheererJ) [inline pkg/gardenadm/cmd/init/init.go:150]: I understand the dependencies in the way that we want to ensure that this is called after the extension namespaces are created. However, I am not sure if this is sufficient. We just wait for the deployment to finish, which essentially creates managed resources. Afterwards, `gardener-resource-manager` would need to actually apply the content.
Do you think that this is a real race condition or just nit-picking?

Second potential issue is that no other task depends on deployment of the network policies. Hence, the flow might run until after `core-dns` deployment and then abort while not yet having deployed the network policies. The next run would try to directly put `gardener-resource-manager` and extensions into the pod network, but without the network policies this will hang.
WDYT?

**Review by ScheererJ — COMMENTED** [review]
Thanks a lot for making autonomous shoot clusters even more usable. The pull request already looks awesome, but I have a few comments.

File: pkg/gardenadm/botanist/network_test.go (hunk 6 above)
Comment (rfranzke) [reply pkg/gardenadm/botanist/network_test.go:79]: Do you have a proposal for writing such a test with the fake client? I'm not sure whether it is reasonable to go down the mock client-path just for this corner case. However, if there is an idea how to realize this with the fake client, I'm happy to add it.

File: pkg/gardenadm/cmd/init/init.go (hunk 9 above)
Comment (rfranzke) [reply pkg/gardenadm/cmd/init/init.go:150]: > I understand the dependencies in the way that we want to ensure that this is called after the extension namespaces are created. However, I am not sure if this is sufficient. We just wait for the deployment to finish, which essentially creates managed resources. Afterwards, `gardener-resource-manager` would need to actually apply the content.
> Do you think that this is a real race condition or just nit-picking?

The namespaces for the extensions are not deployed by `gardener-resource-manager` (via `ManagedResource`) but via the `ControllerInstallation` reconciler. Hence, just depending on the deployment should be enough, no?

File: pkg/gardenadm/botanist/network_test.go (hunk 6 above)
Comment (ScheererJ) [reply pkg/gardenadm/botanist/network_test.go:79]: Clarified in offline discussion. The "easy" approach using a different `scheme` does not seem to work. The additional effort using mock clients does not make sense. Hence, we keep it as is.

File: pkg/gardenadm/cmd/init/init.go (hunk 9 above)
Comment (ScheererJ) [reply pkg/gardenadm/cmd/init/init.go:150]: Yes, that makes sense. Thanks for the clarification.

File: extensions/pkg/controller/controlplane/genericactuator/actuator.go [hunk 10]
```diff
@@ -553,7 +553,7 @@ func (a *actuator) computeChecksums(
 ) (map[string]string, error) {
 	// Get cloud provider secret and config from cluster
 	cpSecret := &corev1.Secret{}
-	if err := a.client.Get(ctx, client.ObjectKey{Namespace: namespace, Name: v1beta1constants.SecretNameCloudProvider}, cpSecret); err != nil {
+	if err := a.client.Get(ctx, client.ObjectKey{Namespace: namespace, Name: v1beta1constants.SecretNameCloudProvider}, cpSecret); client.IgnoreNotFound(err) != nil {
```
Comment (ScheererJ) [inline extensions/pkg/controller/controlplane/genericactuator/actuator.go:556]: Just wondering when looking at the `ConfigMap` code below if we should handle the cloud provider secret in the same way, i.e. have an attribute in the actuator for the secret name. The default could be to have a "cloudprovider" secret, but provider-local could leave this empty to indicate that there is no secret.
This would certainly be the more excessive code change compared to this minimal change, but it would somewhat align the `Secret` and `ConfigMap` handling. WDYT?

File: pkg/gardenadm/botanist/controlplane.go [hunk 11]
```diff
@@ -135,6 +133,25 @@ func (b *AutonomousBotanist) populateStaticAdminTokenToAccessTokenSecret(ctx con
 	return b.SeedClientSet.Client().Update(ctx, secret)
 }
 
+func (b *AutonomousBotanist) filesForStaticControlPlanePods(ctx context.Context) ([]extensionsv1alpha1.File, error) {
+	var files []extensionsv1alpha1.File
+
+	for _, component := range b.staticControlPlaneComponents() {
+		if err := b.SeedClientSet.Client().Get(ctx, client.ObjectKeyFromObject(component.targetObject), component.targetObject); err != nil {
```
Comment (ScheererJ) [inline pkg/gardenadm/botanist/controlplane.go]: How is this working? As far as I can tell, the `targetObject` is just an empty resource without name/namespace being specified.

File: pkg/component/etcd/bootstrap/etcd.go [hunk 12]
```diff
@@ -98,162 +99,171 @@ func (e *etcdDeployer) Deploy(ctx context.Context) error {
...
 					},
 				},
 			},
 		}
 
+		statefulSet.Labels = map[string]string{
+			v1beta1constants.LabelApp:   "etcd",
+			v1beta1constants.LabelRole:  e.values.Role,
+			v1beta1constants.GardenRole: v1beta1constants.GardenRoleControlPlane,
+		}
```
Comment (ScheererJ) [inline pkg/component/etcd/bootstrap/etcd.go]:
```suggestion
		statefulSet.Labels = e.labels()
```

File: pkg/component/etcd/bootstrap/etcd.go [hunk 13]
```diff
@@ -264,6 +274,14 @@ func (e *etcdDeployer) Destroy(_ context.Context) error {
 	return nil
 }
 
-func (e *etcdDeployer) emptyPod() *corev1.Pod {
-	return &corev1.Pod{ObjectMeta: metav1.ObjectMeta{Name: "etcd-" + e.values.Role + "-0", Namespace: e.namespace}}
+func (e *etcdDeployer) emptyStatefulSet() *appsv1.StatefulSet {
+	return &appsv1.StatefulSet{ObjectMeta: metav1.ObjectMeta{Name: "etcd-" + e.values.Role + "-0", Namespace: e.namespace}}
```
Comment (ScheererJ) [inline pkg/component/etcd/bootstrap/etcd.go:273]:
```suggestion
	return &appsv1.StatefulSet{ObjectMeta: metav1.ObjectMeta{Name: "etcd-" + e.values.Role, Namespace: e.namespace}}
```

However, if the suffix is removed some of the replacements `pod.Name` => `statefulSet.Name` will no longer yield the same result.

File: pkg/gardenadm/cmd/init/init.go [hunk 14]
```diff
@@ -232,6 +232,11 @@ func run(ctx context.Context, opts *Options) error {
 			Fn:           b.Shoot.Components.Extensions.ControlPlane.Wait,
 			Dependencies: flow.NewTaskIDs(deployControlPlane),
 		})
+		_ = g.Add(flow.Task{
+			Name:         "Deploying control plane components as Deployments for static pod translation",
+			Fn:           b.DeployControlPlaneDeployments,
+			Dependencies: flow.NewTaskIDs(syncPointBootstrapped),
+		})
```
Comment (ScheererJ) [inline pkg/gardenadm/cmd/init/init.go]: I am a bit confused. Why is this deployed explicitly here, but implicitly as well for the operating system config? I guess for the first deployment of the operating system config we do not expect any interaction, i.e. there are no webhooks, but here we assume that the provider extension could mutate the deployments.
Nevertheless, I propose to make it explicit in both cases, i.e. deploy the control plane deployments twice. The first deployment is an explicit precondition for deploying the operating system config and the second one here is done for webhooks, which should probably also be noted in the name/description.
WDYT?

Comment (rfranzke) [reply pkg/gardenadm/cmd/init/init.go]: In the bootstrap flow, this happens implicitly in

```go
deployOperatingSystemConfigSecretForNodeAgent = g.Add(flow.Task{
	Name:         "Generating OperatingSystemConfig and deploying Secret for gardener-node-agent",
	Fn:           b.DeployOperatingSystemConfigSecretForNodeAgent,
	SkipIf:       kubeconfigFileExists,
	Dependencies: flow.NewTaskIDs(initializeSecretsManagement),
})
```

Personally, I don't think there is much value in running this function explicitly in this flow - nobody really cares about how this works under the hood. This is different later. That's why I decided to do it explicitly only there.

TL;DR: If you insist, I can also make it a dedicated task in the bootstrap flow, but I were to decide, I would keep it this way 😄

File: pkg/component/etcd/bootstrap/etcd.go (hunk 13 above)
Comment (rfranzke) [reply pkg/component/etcd/bootstrap/etcd.go:273]: Why should we remove the suffix? I don't know how this all will work with HA etcd clusters, but I guess we will also end up with three `StatefulSet`s for `etcd-{0,1,2}` that we can then translate and push to the various control plane nodes. With only one `StatefulSet`, this approach won't work.
Hence, I would keep it this way for now until this part becomes more clear. WDYT?

File: extensions/pkg/controller/controlplane/genericactuator/actuator.go (hunk 10 above)
Comment (rfranzke) [reply extensions/pkg/controller/controlplane/genericactuator/actuator.go:556]: > if we should handle the cloud provider secret in the same way, i.e. have an attribute in the actuator for the secret name.

The name of the cloud provider `Secret` is always the same (and dictated/deployed via `gardenlet`). The name of the `ConfigMap`, however, is dictated by the extension itself, i.e., the generic actuator cannot know it.

> but provider-local could leave this empty to indicate that there is no secret.

Hm, this is not about `provider-local`. Currently, we simply don't use/work with a cloud provider secret in the `gardenadm init` flow. This will probably change later when we look at using real provider extensions (or latest when we look at the "medium touch" case), but for now, there is no such task/step necessary.

File: pkg/gardenadm/botanist/controlplane.go (hunk 11 above)
Comment (rfranzke) [reply pkg/gardenadm/botanist/controlplane.go]: You're right, this didn't work, hence, it was adapted in the following commit `Deploy control plane components with 0 replicas` (https://github.com/gardener/gardener/pull/11892/commits/763159712a00fd00f3e91aa023613dc143c581af#diff-7a1471a4fe49e105383d4a81fe92976c516d1c941a553365b6a2361606146dc8R142). Sorry that this was confusing in this intermediate commit.

**Review by ScheererJ — APPROVED** [review]
/lgtm
/approve
